### PR TITLE
[x] Make serialize.TagValueFromEncodedTagsFast() faster

### PR DESCRIPTION
**Pull request by vpranckaitis**
**What this PR does / why we need it**:

Adds a benchmark for `serialize.TagValueFromEncodedTagsFast()` function and optimizes a couple of lines in it.

Before and after function change:

```
BenchmarkTagValueFromEncodedTagsFast-12    	 8657110	       140 ns/op
BenchmarkTagValueFromEncodedTagsFast-12    	11756650	       110 ns/op
```

Spinoff from #3087

**Special notes for your reviewer**:

**Does this PR introduce a user-facing and/or backwards incompatible change?**:

```release-note
NONE
```

**Does this PR require updating code package or user-facing documentation?**:

```documentation-note
NONE
```

### CODE REVIEW

File: src/x/serialize/decoder_fast_benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	encodedTags, err := base64.StdEncoding.DecodeString(samples[0])
+	if err != nil {
+		return nil, err
+	}
+
+	tagNames, err := decodeTagNames(encodedTags)
+	if err != nil {
+		return nil, err
+	}
+	tagNames = append(tagNames, []byte("not_exist"))
```
Comment (linasm) [inline src/x/serialize/decoder_fast_benchmark_test.go:85]: Do I understand it correctly that the benchmark has 20 different tag sets, but the search is always performed for one of the tags from the first tag set only? If that is the case, what is the overlap of tag names between those tag set samples? 
(basically I'm trying to understand if the benchmark is representative enough)

File: src/x/serialize/decoder_fast_benchmark_test.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+func prepareData(b *testing.B) ([]encodedTagsWithTagName, error) {
+	encodedTags, err := base64.StdEncoding.DecodeString(samples[0])
+	if err != nil {
+		return nil, err
+	}
+
+	tagNames, err := decodeTagNames(encodedTags)
+	if err != nil {
+		return nil, err
+	}
```
Comment (linasm) [inline src/x/serialize/decoder_fast_benchmark_test.go:84]: Nit:
```suggestion
	require.NoError(b, err)
```

File: src/x/serialize/decoder_fast_benchmark_test.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+		_, _, err := TagValueFromEncodedTagsFast(testData[i].encodedTags, testData[i].tagName)
+		require.NoError(b, err)
+	}
+}
+
+func prepareData(b *testing.B) ([]encodedTagsWithTagName, error) {
+	encodedTags, err := base64.StdEncoding.DecodeString(samples[0])
+	if err != nil {
+		return nil, err
+	}
```
Comment (linasm) [inline src/x/serialize/decoder_fast_benchmark_test.go:79]: Nit:
```suggestion
	require.NoError(b, err)
```

File: src/x/serialize/decoder_fast_benchmark_test.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	var (
+		result = make([]encodedTagsWithTagName, 0, b.N)
+		rnd    = rand.New(rand.NewSource(42)) //nolint:gosec
+	)
+	for i := 0; i < b.N; i++ {
+		tagName := tagNames[rnd.Intn(len(tagNames))]
+		encodedTags, err = base64.StdEncoding.DecodeString(samples[rnd.Intn(len(samples))])
+		if err != nil {
+			return nil, err
+		}
```
Comment (linasm) [inline src/x/serialize/decoder_fast_benchmark_test.go:96]: Nit:
```suggestion
		require.NoError(b, err)
```
(and then this fn does not have to return an error)

**Review by linasm — COMMENTED** [review]
Some feedback on the benchmark, also a general suggestion regarding the commit message:
I think we want to emphasise what the change/improvement in the production code is, and the benchmark is just a byproduct of that.

File: src/x/serialize/decoder_fast_benchmark_test.go (hunk 3 above)
Comment (vpranckaitis) [reply src/x/serialize/decoder_fast_benchmark_test.go]: https://github.com/m3db/m3/pull/3097/commits/59887f402095dd6a53fa13172565d2d27663d5b4#diff-420a2d9b9e8a8420ae2f3add1a14145260f611efe27a68859cd7272b1971f3bbR76

File: src/x/serialize/decoder_fast_benchmark_test.go (hunk 2 above)
Comment (vpranckaitis) [reply src/x/serialize/decoder_fast_benchmark_test.go]: https://github.com/m3db/m3/pull/3097/commits/59887f402095dd6a53fa13172565d2d27663d5b4#diff-420a2d9b9e8a8420ae2f3add1a14145260f611efe27a68859cd7272b1971f3bbR79

File: src/x/serialize/decoder_fast_benchmark_test.go (hunk 4 above)
Comment (vpranckaitis) [reply src/x/serialize/decoder_fast_benchmark_test.go]: https://github.com/m3db/m3/pull/3097/commits/59887f402095dd6a53fa13172565d2d27663d5b4#diff-420a2d9b9e8a8420ae2f3add1a14145260f611efe27a68859cd7272b1971f3bbR89

File: src/x/serialize/decoder_fast_benchmark_test.go (hunk 1 above)
Comment (vpranckaitis) [reply src/x/serialize/decoder_fast_benchmark_test.go]: Used the data from 

https://github.com/m3db/m3/blob/ed786ba6b6845daaeb106756838c885c854d2864/src/dbnode/storage/index/convert/convert_benchmark_test.go#L52-L131

Each sample has the same set of tags (added a comment https://github.com/m3db/m3/pull/3097/commits/59887f402095dd6a53fa13172565d2d27663d5b4#diff-420a2d9b9e8a8420ae2f3add1a14145260f611efe27a68859cd7272b1971f3bbR77)

**Review by linasm — APPROVED** [review]
LGTM. 
I suggest (as a low priority follow-up task) to do a search for `bytes.Compare` across the code base and make the same improvement where applicable. I see there are at least few more instances where it is being used for equality checks. No need for additional benchmarks as we already saw that this change is beneficial (unless you happen to find some already existing benchmarks for the code that you change).

**Pull request merged**